File: java/src/blackjackgame/BlackJackGame.java
package blackjackgame;

import card.Deck;
import player.Dealer;
import player.AbstractPlayer;
import player.User;

public class BlackJackGame {

    /**
     *  ゲーム開始
     */
    public void start() {
        System.out.println("★☆★☆★☆★☆★☆★☆　ブラックジャックにようこそ！　★☆★☆★☆★☆★☆★☆\n");
        System.out.println("ゲームを開始します。\n");

        Deck deck     = new Deck();
        AbstractPlayer user   = new User("あなた");
        AbstractPlayer dealer = new Dealer("ディーラー");

        user.initCardList(deck);
        dealer.initCardList(deck);

        user.drawCard(deck);
        if(!user.isBust()) dealer.drawCard(deck);

        printGameResult(user, dealer);

        System.out.println("\nブラックジャック終了！また遊んでね★");
    }

    /**
     * ゲームの結果を表示
     * @param player1 プレイヤー1
     * @param player2 プレイヤー2
     */
    private void printGameResult(AbstractPlayer player1, AbstractPlayer player2) {
        if (player1.calcScore() == player2.calcScore()) {
            System.out.println("引き分けです。");
            return;
        }
        AbstractPlayer winner = !player1.isBust() && (player2.isBust() || player1.calcScore() > player2.calcScore())
                ? player1
                : player2;
        System.out.println( winner.getName() + "の勝ちです！" );
    }
}

File: scala/src/main/scala/BlackJackGame.scala
case class BlackJackGame() {
  def start(): Unit = {
    println("★☆★☆★☆★☆★☆★☆　ブラックジャックにようこそ！　★☆★☆★☆★☆★☆★☆\n")
    println("ゲームを開始します。\n")

    val deck: Deck = Deck()
    val user: AbstractPlayer = User("あなた")
    val dealer: AbstractPlayer = Dealer("ディーラー")

    user.initCardList(deck)
    dealer.initCardList(deck)

    user.drawCard(deck)
    if (!user.isBust) {
      dealer.drawCard(deck)
    }

    printGameResult(user, dealer)

    println("\nブラックジャック終了！また遊んでね★")
  }

  def printGameResult(player1: AbstractPlayer, player2: AbstractPlayer): Unit = {
    if (player1.calcScore() == player2.calcScore()) {
      println("引き分けです。")
      return
    }
    val winner: AbstractPlayer = if (!player1.isBust && (player2.isBust || player1.calcScore() > player2.calcScore())) {
      player1
    } else {
      player2
    }
    println(winner.name + "の勝ちです！")
  }
}
